torchao/prototype/mx_formats/custom_cast.py

```diff
@@ -0,0 +1,713 @@
+    # numerical correctness. There is likely a way to optimize the performance,
+    # I just haven't had time to look into it.
+    if lp_dtype_name == DTYPE_FP4:
+        result[denormal_mask] = ZERO_POINT_FIVE_BITS_F32
+
+    elif lp_dtype_name == DTYPE_FP6_E2M3:
+        # Only 7 possible values, just do a LUT
+        # Note: calculate the booleans first because we are modifying
+        # this variable inplace.
+        is_val1 = mantissa_lp_int32 == 1
+        is_val2 = mantissa_lp_int32 == 2
+        is_val3 = mantissa_lp_int32 == 3
+        is_val4 = mantissa_lp_int32 == 4
+        is_val5 = mantissa_lp_int32 == 5
+        is_val6 = mantissa_lp_int32 == 6
+        is_val7 = mantissa_lp_int32 == 7
+        mantissa_lp_int32[is_val1] = 0x3E000000  # 0.125
+        mantissa_lp_int32[is_val2] = 0x3E800000  # 0.25
+        mantissa_lp_int32[is_val3] = 0x3EC00000  # 0.375
+        mantissa_lp_int32[is_val4] = 0x3F000000  # 0.5
+        mantissa_lp_int32[is_val5] = 0x3F200000  # 0.625
+        mantissa_lp_int32[is_val6] = 0x3F400000  # 0.75
+        mantissa_lp_int32[is_val7] = 0x3F600000  # 0.875
+        result = torch.where(denormal_mask, mantissa_lp_int32, result)
+
+    elif lp_dtype_name == DTYPE_FP6_E3M2:
+        # Only 3 possible values, just do a LUT
+        # Note: calculate the booleans first because we are modifying
+        # this variable inplace.
+        is_val1 = mantissa_lp_int32 == 1
+        is_val2 = mantissa_lp_int32 == 2
+        is_val3 = mantissa_lp_int32 == 3
+        mantissa_lp_int32[is_val1] = 0x3D800000  # 0.0625
+        mantissa_lp_int32[is_val2] = 0x3E000000  # 0.125
+        mantissa_lp_int32[is_val3] = 0x3E400000  # 0.1875
+        result = torch.where(denormal_mask, mantissa_lp_int32, result)
+    else:
+        raise AssertionError(f"unsupported lp_dtype_name {lp_dtype_name}")
+
+    # add sign back
+    sign_f32 = sign_lp.to(torch.int32) << (
+        MBITS_F32 - mbits + EBITS_F32 - ebits
+    )  # noqa: E501
+    result = result | sign_f32
+
+    return result.view(torch.float)
+
+
+def f4_unpacked_to_f32(x: torch.Tensor):
+    """
+    Input: torch.Tensor of dtype uint8, with bits 0-3 empty and bits 4-7
+      containing an fp4_e2m1 encoding
+    Output: torch.Tensor of dtype fp32 with the dequantized value
+    """
+    return _f4_or_f6_unpacked_to_f32(x, DTYPE_FP4)
+
+
+def f6_e2m3_unpacked_to_f32(x: torch.Tensor):
+    """
+    Input: torch.Tensor of dtype uint8, with bits 0-1 empty and bits 2-7
+      containing an fp6_e3m2 encoding
+    Output: torch.Tensor of dtype fp32 with the dequantized value
+    """
+    return _f4_or_f6_unpacked_to_f32(x, DTYPE_FP6_E2M3)
+
+
+def f6_e3m2_unpacked_to_f32(x: torch.Tensor):
+    """
+    Input: torch.Tensor of dtype uint8, with bits 0-1 empty and bits 2-7
+      containing an fp6_e3m2 encoding
+    Output: torch.Tensor of dtype fp32 with the dequantized value
+    """
+    return _f4_or_f6_unpacked_to_f32(x, DTYPE_FP6_E3M2)
+
+
+if has_triton():
```

Comment:
was the inductor codegen note adequate? Wondering if we can eventually remove this

Reply:
current codegen was slow, tracked in https://github.com/pytorch/pytorch/issues/124002 .